pallets/proxy-bonding/src/functions.rs

```diff
@@ -0,0 +1,114 @@
+	/// User pays a fee for this functionality, which can be later refunded.
+	pub fn bond_on_behalf_of(
+		derivation_path: u32,
+		account: T::AccountId,
+		bond_amount: BalanceOf<T>,
+		fee_asset: AssetId,
+		hold_reason: T::RuntimeHoldReason,
+	) -> Result<(), DispatchError> {
+		let treasury = T::Treasury::get();
+		let bonding_account: AccountIdOf<T> = T::RootId::get().into_sub_account_truncating(derivation_path);
+		let existential_deposit = <T::BondingToken as fungible::Inspect<T::AccountId>>::minimum_balance();
+
+		let fee_in_fee_asset = Self::calculate_fee(bond_amount, fee_asset)?;
+
+		// Pay the fee from the user to the bonding account. It awaits either a full transfer to the T::FeeRecipient, or a refund to each user
+		T::FeeToken::transfer(fee_asset, &account, &bonding_account, fee_in_fee_asset, Preservation::Preserve)?;
+
+		// Ensure the sub-account has an ED by the treasury. This will be refunded after all the tokens are unlocked
+		if T::BondingToken::balance(&bonding_account) < existential_deposit {
+			T::BondingToken::transfer(
+				&treasury.clone(),
+				&bonding_account,
+				existential_deposit,
+				Preservation::Preserve,
+			)?;
+		}
+		// Bond the PLMC on behalf of the user
+		T::BondingToken::transfer_and_hold(
+			&hold_reason.into(),
+			&treasury.clone(),
+			&bonding_account.clone(),
+			bond_amount,
+			Precision::Exact,
+			Preservation::Preserve,
+			Fortitude::Polite,
+		)?;
+
+		Ok(())
+	}
+
+	/// Set the block for which we can release the bonds of a sub-account, and transfer it back to the treasury.
+	pub fn set_release_type(
+		derivation_path: u32,
+		hold_reason: T::RuntimeHoldReason,
+		release_type: ReleaseType<BlockNumberFor<T>>,
+	) {
+		Releases::<T>::insert(derivation_path, hold_reason, release_type);
+	}
+
+	/// Refund the fee paid by a user to lock up some treasury tokens. It is this function's caller responsibility to ensure that the fee is not refunded twice.
+	pub fn refund_fee(
+		derivation_path: u32,
+		hold_reason: T::RuntimeHoldReason,
+		account: &T::AccountId,
+		bond_amount: BalanceOf<T>,
+		fee_asset: AssetId,
+	) -> Result<(), DispatchError> {
+		let bonding_account: AccountIdOf<T> = T::RootId::get().into_sub_account_truncating(derivation_path);
+		let fee_in_fee_asset = Self::calculate_fee(bond_amount, fee_asset)?;
+		let release_type = Releases::<T>::get(derivation_path, hold_reason).ok_or(Error::<T>::ReleaseTypeNotSet)?;
+
+		ensure!(release_type == ReleaseType::Refunded, Error::<T>::FeeRefundDisallowed);
+
+		// We know this fee token account is existing thanks to the provider reference of the ED of the native asset, so we can fully move all the funds.
+		// FYI same cannot be said of the `account`. We assume they only hold the fee token so their fee asset balance must not go below the min_balance.
+		T::FeeToken::transfer(fee_asset, &bonding_account, account, fee_in_fee_asset, Preservation::Expendable)?;
```

Comment:
Not sure if this will now fail if `fee_in_fee_asset` is below minimum amount for `account`. If its dusted or if the transaction fails. We could already prevent any problems if we call this `refund_fee` after `release_funding_asset` as we are sure that the contribution amount is already enough.

Reply:
Good point, we know for sure the funding ticket is > ED, but we have no guarantees the OTM fee is.
FYI This will be changed in #401